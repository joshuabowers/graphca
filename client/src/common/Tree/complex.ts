import { Complex } from './Expression';
// import { Complex } from './Constant/Complex'

export function complex(a: number, b: number) {
  return new Complex(a, b)
}
