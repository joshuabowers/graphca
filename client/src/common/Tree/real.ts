import { Real } from "./Expression";
// import { Real } from './Constant/Real'

export function real(value: number) {
  return new Real(value)
}
