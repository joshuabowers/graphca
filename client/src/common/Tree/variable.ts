import { Base } from './Expression'

export class Variable extends Base {
  readonly $kind = 'Variable'
  constructor(readonly name: string, readonly value?: Base) { super() }
}

export function variable(name: string, value?: Base) {
  return new Variable(name, value)
}
