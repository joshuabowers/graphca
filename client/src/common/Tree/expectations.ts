import { method, multi, Multi, _ } from '@arrows/multimethod'
import { is } from './is'
import { Real } from './real'
import { Complex } from './complex'

export type ExpectCloseTo = Multi 
  & ((actual: Real, expected: Real, precision: number) => void)
  & ((actual: Complex, expected: Complex, precision: number) => void)

export const expectCloseTo: ExpectCloseTo = multi(
  method(
    [is(Real), is(Real), _], 
    (actual: Real, expected: Real, precision: number) => 
      expect(actual.value).toBeCloseTo(expected.value, precision)
  ),
  method(
    [is(Complex), is(Complex), _],
    (actual: Complex, expected: Complex, precision: number) => {
      expect(actual.a).toBeCloseTo(expected.a, precision)
      expect(actual.b).toBeCloseTo(expected.b, precision)
    }
  )
)
