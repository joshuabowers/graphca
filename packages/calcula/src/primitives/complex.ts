import { Species } from "../utility/tree"
import { Complex, primitive } from "../closures/primitive"
import { isNumberTuple } from "../utility/valuePredicates"
export { Complex }

export const [complex, isComplex, $complex] = 
  primitive<[number, number], {a: number, b: number}, Complex>(
  isNumberTuple,
  ([a, b]) => ({a, b}),
  Species.complex
)(
  r => [r.value, 0],
  c => [c.a, c.b],
  b => [b.value ? 1 : 0, 0]
)()

export const ComplexInfinity = complex([Infinity, NaN])
