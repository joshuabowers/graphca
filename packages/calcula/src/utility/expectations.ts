import { method, multi, Multi, _ } from '@arrows/multimethod'
import { isWriter, Writer } from '../monads/writer'
import { TreeNode } from './tree'
import { Real, Complex, isReal, isComplex } from '../primitives'

export type ExpectCloseTo = Multi 
  & ((actual: Writer<Real>, expected: Writer<Real>, precision: number) => void)
  & ((actual: Writer<Complex>, expected: Writer<Complex>, precision: number) => void)

export const expectCloseTo: ExpectCloseTo = multi(
  method(
    [isReal, isReal, _], 
    (actual: Writer<Real>, expected: Writer<Real>, precision: number) => 
      expect(actual.value.value).toBeCloseTo(expected.value.value, precision)
  ),
  method(
    [isComplex, isComplex, _],
    (actual: Writer<Complex>, expected: Writer<Complex>, precision: number) => {
      expect(actual.value.a).toBeCloseTo(expected.value.a, precision)
      expect(actual.value.b).toBeCloseTo(expected.value.b, precision)
    }
  )
)

type Input = TreeNode|Writer<TreeNode>|[Input, Input]
export type Operation = [Input, string]

export const expectWriter = <
  Actual extends TreeNode, 
  Expected extends TreeNode|Writer<TreeNode>
>(
  actual: Writer<Actual>
) => (expected: Expected, ...operations: Operation[]) => {
  expect(actual).toEqual({
    value: isWriter(expected) ? expected.value : expected,
    log: operations.map(([input, action]) => ({
      input: isWriter(input) 
        ? input.value 
        : (Array.isArray(input) 
          ? input.map(i => isWriter(i) ? i.value : i)
          : input), 
      action
    }))
  })
}
