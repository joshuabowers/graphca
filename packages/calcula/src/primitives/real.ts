import { Species } from '../utility/tree'
import { Real, primitive } from '../closures/primitive'
import { isNumericOrString } from '../utility/valuePredicates'
import { numeric } from '../utility/numeric'
import { nan } from './nan'
export { Real }

export const [real, isReal, $real] = 
  primitive<[number|string], number, Real>(
  isNumericOrString,
  value => Number(value),
  Species.real,
  r => numeric(r.raw)
)(
  r => [r.raw],
  c => [c.raw.a],
  b => [b.raw ? 1 : 0]
)(
  when => [
    when(v => Number.isNaN(v), [nan, 'incalculable'])
  ]
)

export const EulerMascheroni = real(0.57721566490153286060)
