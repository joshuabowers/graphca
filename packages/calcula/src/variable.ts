import { Writer, unit } from './monads/writer'
import { TreeNode, Clades, Species, isSpecies } from './utility/tree'
import { isNil, nil } from './primitives/nil'
import { rule, process, resolve } from './utility/rule'

export type Variable = TreeNode & {
  readonly clade: Clades.variadic,
  readonly species: Species.variable,
  readonly name: string,
  readonly value: Writer<TreeNode>
}

export const isVariable = isSpecies<Variable>(Species.variable)

export const variable = (
  name: string, value: Writer<TreeNode> = nil
): Writer<Variable> => 
  ({
    value: {
      clade: Clades.variadic,
      species: Species.variable,
      name, value
    },
    log: [{
      input: process`${name}`, 
      rewrite: isNil(value) ? resolve`${name}` : process`${value}`,
      action: isNil(value) 
        ? 'referenced variable' : 'referenced variable with value'
    }] 
  })

export type Scope = Map<string, Writer<Variable>>

export const isScope = (v: unknown): v is Scope =>
  typeof v === 'object' && v !== null
  && ['get', 'set', 'has', 'delete'].every(m => m in v)

type Entries = Iterable<readonly [string, Writer<Variable>]>

export const scope = (entries: Entries = []): Scope => 
  new Map<string, Writer<Variable>>(entries)

export const assign = (name: string, value: Writer<TreeNode>, scope?: Scope) => {
  const v = variable(name, value)
  if(scope){
    if(isNil(value)){ scope.delete(name) }
    else { scope.set(name, v) }
  }
  return v
}
