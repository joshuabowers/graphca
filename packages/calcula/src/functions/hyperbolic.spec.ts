import { real } from './real'
import { complex } from './complex'
import { variable } from './variable'
import { 
  HyperbolicCosine, HyperbolicSine, HyperbolicTangent,
  HyperbolicSecant, HyperbolicCosecant, HyperbolicCotangent,
  cosh, sinh, tanh, sech, csch, coth 
} from './hyperbolic'
import { expectCloseTo } from './expectations'

describe('cosh', () => {
  it('calculates the hyperbolic cosine of a real value', () => {
    expectCloseTo(cosh(real(1)), real(1.543080634815), 10)
  })
  
  it('calculates the hyperbolic cosine of a complex number', () => {
    expectCloseTo(cosh(complex(0, 1)), complex(0.540302305868, 0), 10)
  })

  it('returns a hyperbolic cosine node if not valuable', () => {
    expect(cosh(variable('x'))).toEqual(new HyperbolicCosine(variable('x')))
  })
})

describe('sinh', () => {
  it('calculates the hyperbolic sine of a real value', () => {
    expectCloseTo(sinh(real(1)), real(1.175201193643), 10)
  })

  it('calculates the hyperbolic sine of a complex number', () => {
    expectCloseTo(sinh(complex(0, 1)), complex(0, 0.841470984807), 10)
  })

  it('returns a hyperbolic sine node if not valuable', () => {
    expect(sinh(variable('x'))).toEqual(new HyperbolicSine(variable('x')))
  })
})

describe('tanh', () => {
  it('calculates the hyperbolic tangent of a real value', () => {
    expectCloseTo(tanh(real(1)), real(0.761594155955), 10)
  })

  it('calculates the hyperbolic tangent of a complex number', () => {
    expectCloseTo(tanh(complex(0, 1)), complex(0, 1.557407724654), 10)
  })

  it('returns a hyperbolic tangent node if not valuable', () => {
    expect(tanh(variable('x'))).toEqual(new HyperbolicTangent(variable('x')))
  })
})

describe('sech', () => {
  it('calculates the hyperbolic secant of a real value', () => {
    expectCloseTo(sech(real(1)), real(0.648054273663), 10)
  })

  it('calculates the hyperbolic secant of a complex number', () => {
    expectCloseTo(sech(complex(0, 1)), complex(1.850815717680, 0), 10)
  })

  it('returns a hyperbolic secant node if not valuable', () => {
    expect(sech(variable('x'))).toEqual(new HyperbolicSecant(variable('x')))
  })
})

describe('csch', () => {
  it('calculates the hyperbolic cosecant of a real value', () => {
    expectCloseTo(csch(real(1)), real(0.850918128239), 10)
  })

  it('calculates the hyperbolic cosecant of a complex number', () => {
    expectCloseTo(csch(complex(0, 1)), complex(0, -1.188395105778), 10)
  })

  it('returns a hyperbolic cosecant node if not valuable', () => {
    expect(csch(variable('x'))).toEqual(new HyperbolicCosecant(variable('x')))
  })
})

describe('coth', () => {
  it('calculates the hyperbolic cotangent of a real value', () => {
    expectCloseTo(coth(real(1)), real(1.313035285499), 10)
  })

  it('calculates the hyperbolic cotangent of a complex number', () => {
    expectCloseTo(coth(complex(0, 1)), complex(0, -0.642092615934), 10)
  })

  it('returns a hyperbolic cotangent node if not valuable', () => {
    expect(coth(variable('x'))).toEqual(new HyperbolicCotangent(variable('x')))
  })
})
