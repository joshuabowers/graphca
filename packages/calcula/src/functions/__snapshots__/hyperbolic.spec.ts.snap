// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cosh returns a hyperbolic cosine node if not valuable 1`] = `
[
  [
    "cosh(x)",
    "identified hyperbolic cosine",
  ],
  [
    "cosh([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "cosh({x})",
    "processed argument",
  ],
  [
    "cosh(x)",
    "created hyperbolic cosine",
  ],
]
`;

exports[`coth returns a hyperbolic cotangent node if not valuable 1`] = `
[
  [
    "coth(x)",
    "identified hyperbolic cotangent",
  ],
  [
    "coth([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "coth({x})",
    "processed argument",
  ],
  [
    "coth(x)",
    "created hyperbolic cotangent",
  ],
]
`;

exports[`csch returns a hyperbolic cosecant node if not valuable 1`] = `
[
  [
    "csch(x)",
    "identified hyperbolic cosecant",
  ],
  [
    "csch([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "csch({x})",
    "processed argument",
  ],
  [
    "csch(x)",
    "created hyperbolic cosecant",
  ],
]
`;

exports[`sech calculates the hyperbolic secant of a real value 1`] = `
[
  [
    "sech(1)",
    "identified hyperbolic secant",
  ],
  [
    "sech([1])",
    "processing argument",
  ],
  [
    "1",
    "created real",
  ],
  [
    "sech({1})",
    "processed argument",
  ],
  [
    "sech(1)",
    "real hyperbolic secant",
  ],
  [
    "(cosh(1)^-1)",
    "identified exponentiation",
  ],
  [
    "([cosh(1)]^-1)",
    "processing left operand",
  ],
  [
    "cosh(1)",
    "identified hyperbolic cosine",
  ],
  [
    "cosh([1])",
    "processing argument",
  ],
  [
    "1",
    "created real",
  ],
  [
    "cosh({1})",
    "processed argument",
  ],
  [
    "cosh(1)",
    "real hyperbolic cosine",
  ],
  [
    "1.5430806348152437",
    "created real",
  ],
  [
    "({1.5430806348152437}^[-1])",
    "processed left operand; processing right operand",
  ],
  [
    "-1",
    "created real",
  ],
  [
    "(1.5430806348152437^{-1})",
    "processed right operand",
  ],
  [
    "(1.5430806348152437^-1)",
    "real exponentiation",
  ],
  [
    "0.6480542736638855",
    "created real",
  ],
]
`;

exports[`sech returns a hyperbolic secant node if not valuable 1`] = `
[
  [
    "sech(x)",
    "identified hyperbolic secant",
  ],
  [
    "sech([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "sech({x})",
    "processed argument",
  ],
  [
    "sech(x)",
    "created hyperbolic secant",
  ],
]
`;

exports[`sinh returns a hyperbolic sine node if not valuable 1`] = `
[
  [
    "sinh(x)",
    "identified hyperbolic sine",
  ],
  [
    "sinh([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "sinh({x})",
    "processed argument",
  ],
  [
    "sinh(x)",
    "created hyperbolic sine",
  ],
]
`;

exports[`tanh returns a hyperbolic tangent node if not valuable 1`] = `
[
  [
    "tanh(x)",
    "identified hyperbolic tangent",
  ],
  [
    "tanh([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "tanh({x})",
    "processed argument",
  ],
  [
    "tanh(x)",
    "created hyperbolic tangent",
  ],
]
`;
